rfcs/text/0005_route_handler.md

```diff
@@ -0,0 +1,184 @@
+}
+router.get(...,
+  (context: Context, request: KibanaRequest, t: KibanaResponseToolkit): KibanaResponse => {
+   return t.ok();
+   // or
+   return t.redirected('/url');
+   // or
+   return t.badRequest(error);
+);
+```
+*KibanaResponseToolkit*  methods allow an end user to adjust the next response parameters:
+- Body. Supported values:`undefined | string | JSONValue | Buffer | Stream`.
+- Status code. 
+- Headers. Supports adjusting [known values](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v10/http.d.ts#L8) and attaching [custom values as well](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v10/http.d.ts#L67)
+
+Other response parameters, such as `etag`, `MIME-type`, `bytes` that used in the Legacy platform could be adjusted via Headers.
+
+The router handler doesn't expect that logic inside can throw or return something different from `KibanaResponse`. In this case, Http service will respond with `Server error` to prevent exposure of internal logic details.
+
+#### KibanaResponseToolkit methods
+Basic primitives:
+```typescript
+type HttpResponsePayload = undefined | string | JSONValue | Buffer | Stream;
+interface HttpResponseOptions {
+  headers?: {
+   // list of  known headers
+   ...
+   // for custom headers:
+   [header: string]: number | string | string[] | undefined;
+  }
+}
+
+type StatusCode = 200 | 202 | 204 | 302 | 400 | 401 | 403 | 404 | 409 | 500;
+```
+
+##### Success
+Server indicated that request was accepted:
+```typescript
+type SuccessResponse<T> = <T extends HttpResponsePayload>(
+  payload: T,
+  options?: HttpResponseOptions
+) => KibanaResponse<T>;
+
+const kibanaResponseToolkit = {
+  ok: <T extends HttpResponsePayload>(payload: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(200, payload, options),
+  accepted: <T extends HttpResponsePayload>(payload: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(202, payload, options),
+  noContent: (options?: HttpResponseOptions) => new KibanaResponse(204, undefined, options)
+```
+
+##### Redirection
+The server wants a user to perform additional actions.
+```typescript
+const kibanaResponseToolkit = {
+  redirected: (url: string, options?: HttpResponseOptions) => new KibanaResponse(302, url, options),
+  notModified: (options?: HttpResponseOptions) => new KibanaResponse(304, undefined, options),
+```
+
+#####  Error
+Server signals that request cannot be handled and explains details of the error situation
+```typescript
+// Supports attaching additional data to send to the client
+interface ResponseError extends Error {
+  meta: {
+    data?: JSONValue;
+    code?: string; // error code to simplify search, translations in i18n, etc.
+    link?: string; // link to the docs
+  }
+}
+
+const kibanaResponseToolkit = {
+  // Client errors
+  badRequest: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(400, err, options),
```

Comment:
I know `Error` is an interface, but it's also a class which makes creating instances easy. Maybe there should be a convenience method for constructing instances of `ResponseError`.

Reply:
you mean expose explicitly?
```typescript
interface ResponseErrorType extends Error {
  meta?: {
    data?: JSONValue;
    errorCode?: string; // error code to simplify search, translations in i18n, etc.
    docLink?: string; // link to the docs
  };
}

class ResponseError extends Error implements ResponseErrorType {
  constructor(error: Error | string, public readonly meta: ResponseErrorType['meta']) {
    super(typeof error === 'string' ? error : error.message);
    Object.setPrototypeOf(this, ResponseError.prototype);
  }
}

// or a factory
const createError = (error: Error | string, public readonly meta: ResponseErrorType['meta']) => 
  new ResponseError(error, meta)
```

Reply:
Factory makes sense to me, seems more ergonomic / similar to the response factory methods.